Cargo.toml

```diff
@@ -46,6 +46,8 @@ once_cell = "1.3"
 
 cfg-if = "0.1"
 
+turtle_docs_helper = { path = "turtle_docs_helper", optional = true }
```

Comment:
I have tried this out in the latest commit but can't seem to get it too work as it can't seem to find the crate when I import it. (I added some additional info to the commit message as well)

Reply:
Sorry for the trouble! I pulled the code locally and was able to get it working. Here's what you need in your Cargo.toml file:

```toml
[dev-dependencies]
bitvec = "0.17"
turtle_docs_helper = { path = "turtle_docs_helper" }
```

Note that the dependency is not optional since dev-dependencies are apparently not allowed to be optional (that's fine). I unfortunately wasn't able to get `target.'cfg(docs_images)'.dev-dependencies` to work. It seems like it only works for selecting platforms. :(

To run this, I needed to use the command:

```
RUSTDOCFLAGS="--cfg docs_image" cargo test --features "test unstable" --doc
```

A couple of things to note about this:

* using `cargo test` instead of `cargo doc` to make sure the tests are actually run (instead of just generating documentation)
* using the `test` and `unstable` features to make sure none of the tests in the code are left out
* using `--doc` to make sure only doc tests are run since regular unit tests don't generate images

I think the reason you ran into so many issues was because I pointed you to CONTRIBUTING.md which uses `cargo doc`. I don't think there is any way to use a dev-dependency from `cargo doc`. Sorry about that!

Hopefully this resolves the issues you were running into. Don't hesitate to let me know if you have any other questions! :)

Reply:
Sweet, that was going to be my default choice if `target.'cfg(docs_images)'.dev-dependencies` couldn't work :) I think the main difficulty was the first time playing around with cfg to this extent.

Reply:
After implementing the dev-dependancy chance I ran into another issue, the `SaveSvg` wasn't being implemented for the the `Turtle` struct. From testing out a fresh project I haven't been able to get the cfg flags to work for implementing traits for structs. As it won't run the implment block of code it seems. Here is the example I created:

```
// lib.rs for a new crate named "example"
pub struct Cats{}

#[cfg(docs_images)]
impl Cats{
    pub fn say(){
        println!("meow");
    }
}

///
/// ```rust
/// use example::Cats;
/// # #[cfg(docs_images)] Cats::say();
/// # #[cfg(docs_images)] assert_eq!(1,2);
/// 
/// ```
fn foo(){
    println!("hello world");
}
```

It should run the first test then fail the assert but the compiler can't find says(). (This is ran by using cargo test)

Potentailly it's the end of the week and I am missing something obivious as you said you were able to get it working.

Reply:
Could you tell me the command you used to run the program and paste the error message?

Reply:
```
running 1 test
test src\turtle.rs - turtle::Turtle::clear (line 817) ... FAILED

failures:

---- src\turtle.rs - turtle::Turtle::clear (line 817) stdout ----
error[E0277]: the trait bound `turtle::Turtle: turtle_docs_helper::SaveSvg` is not satisfied
  --> src\turtle.rs:824:58
   |
10 | #[cfg(docs_images)] turtle_docs_helper::save_docs_images(&turtle, "clear_before_click");
   |                                                          ^^^^^^^ the trait `turtle_docs_helper::SaveSvg` is not implemented for `turtle::Turtle`
   |
  ::: C:\Users\Joe\workspace\turtle\turtle_docs_helper\src\lib.rs:12:28
   |
12 | pub fn save_docs_images<T: SaveSvg>(drawing: &T, output_name: &str) {
   |                            ------- required by this bound in `turtle_docs_helper::save_docs_images`

error[E0277]: the trait bound `turtle::Turtle: turtle_docs_helper::SaveSvg` is not satisfied
  --> src\turtle.rs:827:58
   |
13 | #[cfg(docs_images)] turtle_docs_helper::save_docs_images(&turtle, "clear_after_click");
   |                                                          ^^^^^^^ the trait `turtle_docs_helper::SaveSvg` is not implemented for `turtle::Turtle`
   |
  ::: C:\Users\Joe\workspace\turtle\turtle_docs_helper\src\lib.rs:12:28
   |
12 | pub fn save_docs_images<T: SaveSvg>(drawing: &T, output_name: &str) {
   |                            ------- required by this bound in `turtle_docs_helper::save_docs_images`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
Couldn't compile the test.

failures:
    src\turtle.rs - turtle::Turtle::clear (line 817)

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 146 filtered out

error: test failed, to rerun pass '--doc'
PS C:\Users\Joe\workspace\turtle> cargo test --features "test unstable" --doc --package turtle -- turtle::Turtle::clear
```
with `RUSTDOCFLAGS` set to `--cfg docs_images` in the `env` 

I can push the code as well although there haven't really be any major changes.

Reply:
> perhaps as I am running as a docs flag it is only affecting the doctests and nothing else

That is possible! Totally not your fault. I would personally expect it to work the way we have it. You can try fixing it by setting `RUSTFLAGS` in addition to `RUSTDOCFLAGS`. Let me know if it works!

And thanks for taking the time to experiment!

Reply:
Yep because doc tests run kind of like integeration tests that was the problem. As the code it was running on didn't get the cfg flag passed to it. For `turtle_docs_helper` to work as a crate we need it to only be required when running tests so: 
```
#[cfg(all(tests, docs_images))]
use turtle_docs_helper;
```
As its a dev dependency so not found when running the main build (`cargo build` ). So that should fix everything but we will see if anything else explodes when I next work on it haha

Reply:
Basically I think a feature flag like orginal commit is the only way to get this work and the nicest way I think. Writing the code is a bit more verbose but it's really simple to run and actually runs.

One of the issues is in order for cfg flags to work they are seperate for `RUSTFLAGS` and `RUSTDOCFLAGS` so that means you need both set if you want the doctests to run the `cfg(docs_images)`.

I think because of that I didn't notice that this:
```

#[cfg(all(test, docs_images))]
use turtle_docs_helper;


pub struct Cats{

}

#[cfg(all(test, docs_images))]
impl Cats{
    pub fn say(){
        println!("meow");
        turtle_docs_helper::cats();
    }
}

///
/// ```rust
/// use example::Cats;
/// # #[cfg(docs_images)] Cats::say();
/// # #[cfg(docs_images)] assert_eq!(turtle_docs_helper::cats(), 4);
/// # #[cfg(docs_images)] assert_eq!(1,2);
/// # #[cfg(docs_images)] assert_eq!(3,5);
/// assert_eq!(5,2);
/// 
/// ```
fn foo(){
    println!("hello world");
}
```

Doesn't work as this won't run (for `test` or `tests`):
```
#[cfg(all(test, docs_images))]
impl Cats{
    pub fn say(){
        println!("meow");
        turtle_docs_helper::cats();
    }
}
```

 The `use turtle_docs_helper` will run so that module does get imported however you get the error:
```
---- src\lib.rs - foo (line 18) stdout ----
error[E0599]: no function or associated item named `say` found for struct `example::Cats` in the current scope
 --> src\lib.rs:20:27
  |
5 | #[cfg(docs_images)] Cats::say();
  |                           ^^^ function or associated item not found in `example::Cats`

error: aborting due to previous error
```